# Capstone 1 — Session 1: Data Import and Cleaning

**Run timestamp:** `2026-02-19 01:49:22`

## Goal
- I inspected and validated the NSMES1988 dataset, documented structure/quality findings, and prepared the cleaned handoff file for the next capstone.
- I also generated required export artifacts and captured evidence for rows, columns, dtypes, missingness, value ranges, and memory usage.

## Inputs
- `NSMES1988.csv`

## Outputs
- All exports go to `./outputs/` (and plots to `./outputs/plots/` when applicable)

## Libraries (documented)
- `pandas`: I used it for loading, profiling, validation checks, and CSV/JSON exports.

## Key dataset note
- `age` is encoded as **Age in years (divided by 10)** (e.g., `6.9` = 69 years).

## C?-T0 — Runtime setup (paths + output folders)
I initialized paths and output folders in this setup block to keep the workflow reproducible.

In [1]:
from pathlib import Path
from datetime import datetime

try:
    from IPython.display import display
except Exception:
    def display(value):
        print(value)

# --- Project metadata ---
CAPSTONE = 1
SESSION_TITLE = 'Session 1: Data Import and Cleaning'

print(f"Capstone: {CAPSTONE} | Session: {SESSION_TITLE}")
print("Run timestamp:", datetime.now().strftime("%Y-%m-%d %H:%M:%S"))

# --- Portable, project-relative paths ---
CWD = Path.cwd()

if CWD.name == f"Capstone {CAPSTONE}":
    BASE_DIR = CWD
elif (CWD / "Incremental_Capstone" / f"Capstone {CAPSTONE}").exists():
    BASE_DIR = CWD / "Incremental_Capstone" / f"Capstone {CAPSTONE}"
elif CWD.parent.name == "Incremental_Capstone" and CWD.name.startswith("Capstone "):
    BASE_DIR = CWD
elif (CWD.parent / "Incremental_Capstone" / f"Capstone {CAPSTONE}").exists():
    BASE_DIR = CWD.parent / "Incremental_Capstone" / f"Capstone {CAPSTONE}"
else:
    BASE_DIR = CWD

DATASET_PATH = BASE_DIR / "NSMES1988.csv"
OUTPUT_DIR = BASE_DIR / "outputs"
PLOTS_DIR = OUTPUT_DIR / "plots"

OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
PLOTS_DIR.mkdir(parents=True, exist_ok=True)

if not DATASET_PATH.exists():
    raise FileNotFoundError(
        f"Dataset not found. Expected at: {DATASET_PATH}. Run notebook from project root or Capstone 1 folder."
    )

print("Base directory:", BASE_DIR)
print("Dataset path:", DATASET_PATH)
print("Output directory:", OUTPUT_DIR)
print("Plots directory:", PLOTS_DIR)

Capstone: 1 | Session: Session 1: Data Import and Cleaning
Run timestamp: 2026-02-19 00:30:45
Base directory: c:\DEV_Projects\SIMPLILEARN\CAPSTONE_Applied_Data_Science_with _Python\Incremental_Capstone\Capstone 1
Dataset path: c:\DEV_Projects\SIMPLILEARN\CAPSTONE_Applied_Data_Science_with _Python\Incremental_Capstone\Capstone 1\NSMES1988.csv
Output directory: c:\DEV_Projects\SIMPLILEARN\CAPSTONE_Applied_Data_Science_with _Python\Incremental_Capstone\Capstone 1\outputs
Plots directory: c:\DEV_Projects\SIMPLILEARN\CAPSTONE_Applied_Data_Science_with _Python\Incremental_Capstone\Capstone 1\outputs\plots


## C?-T1 — Imports I used
I documented each import with why I used it and what it enabled in the analysis.

In [2]:
import pandas as pd  # DataFrames + CSV/JSON IO + analysis tables

## C?-T2 — Load dataset
I loaded the capstone dataset using the required default filename.

In [3]:
DEFAULT_DATASET = "NSMES1988.csv"

if "pd" not in globals():
    import pandas as pd

if "display" not in globals():
    try:
        from IPython.display import display
    except Exception:
        def display(value):
            print(value)

if "Path" not in globals():
    from pathlib import Path

if "BASE_DIR" not in globals() or "DATASET_PATH" not in globals():
    CWD = Path.cwd()
    CAPSTONE = 1

    if CWD.name == f"Capstone {CAPSTONE}":
        BASE_DIR = CWD
    elif (CWD / "Incremental_Capstone" / f"Capstone {CAPSTONE}").exists():
        BASE_DIR = CWD / "Incremental_Capstone" / f"Capstone {CAPSTONE}"
    elif CWD.parent.name == "Incremental_Capstone" and CWD.name.startswith("Capstone "):
        BASE_DIR = CWD
    elif (CWD.parent / "Incremental_Capstone" / f"Capstone {CAPSTONE}").exists():
        BASE_DIR = CWD.parent / "Incremental_Capstone" / f"Capstone {CAPSTONE}"
    else:
        BASE_DIR = CWD

    DATASET_PATH = BASE_DIR / DEFAULT_DATASET

if not DATASET_PATH.exists():
    raise FileNotFoundError(
        f"Dataset not found. Expected at: {DATASET_PATH}. Run notebook from project root or Capstone 1 folder."
    )

dataset_path = DATASET_PATH
df = pd.read_csv(dataset_path)

print("Loaded:", dataset_path)
print("Shape:", df.shape)
display(df.head())

Loaded: c:\DEV_Projects\SIMPLILEARN\CAPSTONE_Applied_Data_Science_with _Python\Incremental_Capstone\Capstone 1\NSMES1988.csv
Shape: (4406, 19)


,Unnamed: 0,visits,nvisits,ovisits,novisits,emergency,hospital,health,chronic,adl,region,age,gender,married,school,income,employed,insurance,medicaid
0,1,5,0,0,0,0,1,average,2,normal,other,6.9,male,yes,6,2.8810,yes,yes,no
1,2,1,0,2,0,2,0,average,2,normal,other,7.4,female,yes,10,2.7478,no,yes,no
2,3,13,0,0,0,3,3,poor,4,limited,other,6.6,female,no,10,0.6532,no,no,yes
3,4,16,0,5,0,1,1,poor,2,limited,other,7.6,male,yes,3,0.6588,no,yes,no
4,5,3,0,0,0,0,0,average,2,limited,other,7.9,female,yes,6,0.6588,no,yes,no


## C?-T3 — Validation checks
- I confirmed expected columns exist
- I confirmed key dtypes
- I checked missing values

**Results Capture:**
- What I did: I validated expected columns, reviewed dtypes, and computed missing-value counts.
- What I found: the raw dataset has shape `(4406, 19)`, all required columns are present, and missing values are zero across all fields.
- Caveats: `Unnamed: 0` behaves like an index-like column and was removed in the cleaned export.

In [4]:
expected_cols = [
    "Unnamed: 0", "visits", "nvisits", "ovisits", "novisits", "emergency", "hospital",
    "health", "chronic", "adl", "region", "age", "gender", "married",
    "school", "income", "employed", "insurance", "medicaid"
]

missing_cols = [c for c in expected_cols if c not in df.columns]
print("Missing expected columns:", missing_cols)

print("\nDtypes:")
display(df.dtypes)

print("\nMissing values (count):")
na_counts = df.isna().sum().sort_values(ascending=False)
display(na_counts[na_counts > 0] if (na_counts > 0).any() else na_counts.head())


Missing expected columns: []

Dtypes:


Unnamed: 0      int64
visits          int64
nvisits         int64
ovisits         int64
novisits        int64
emergency       int64
hospital        int64
health            str
chronic         int64
adl               str
region            str
age           float64
gender            str
married           str
school          int64
income        float64
employed          str
insurance         str
medicaid          str
dtype: object


Missing values (count):


Unnamed: 0    0
visits        0
nvisits       0
ovisits       0
novisits      0
dtype: int64

## C1-T4 — Inspection (rows/cols/dtypes) + basic profiling
**PDF requirement:** Inspect the data and report details such as rows, columns, and data types.

### What I completed
- I profiled the dataset using `shape`, `columns`, `dtypes`, `info`, and numeric summaries.

### Results Capture
- `df.shape = (4406, 19)`.
- I identified 11 numeric and 8 non-numeric columns.
- I flagged `Unnamed: 0` as index-like and removed it in the final cleaned export.

### Code evidence
- The next cell shows the exact profiling code I ran.

In [5]:
# Inspection
print("Shape:", df.shape)
print("\nColumns:")
print(df.columns.tolist())

print("\nInfo:")
df.info()

print("\nDescribe (numeric):")
display(df.describe())

print("\nHead:")
display(df.head())


Shape: (4406, 19)

Columns:
['Unnamed: 0', 'visits', 'nvisits', 'ovisits', 'novisits', 'emergency', 'hospital', 'health', 'chronic', 'adl', 'region', 'age', 'gender', 'married', 'school', 'income', 'employed', 'insurance', 'medicaid']

Info:
<class 'pandas.DataFrame'>
RangeIndex: 4406 entries, 0 to 4405
Data columns (total 19 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  4406 non-null   int64  
 1   visits      4406 non-null   int64  
 2   nvisits     4406 non-null   int64  
 3   ovisits     4406 non-null   int64  
 4   novisits    4406 non-null   int64  
 5   emergency   4406 non-null   int64  
 6   hospital    4406 non-null   int64  
 7   health      4406 non-null   str    
 8   chronic     4406 non-null   int64  
 9   adl         4406 non-null   str    
 10  region      4406 non-null   str    
 11  age         4406 non-null   float64
 12  gender      4406 non-null   str    
 13  married     4406 non-null   str    
 14  s

,Unnamed: 0,visits,nvisits,ovisits,novisits,emergency,hospital,chronic,age,school,income
count,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000
mean,2203.500000,5.774399,1.618021,0.750794,0.536087,0.263504,0.295960,1.541988,7.402406,10.290286,2.527132
std,1272.046972,6.759225,5.317056,3.652759,3.879506,0.703659,0.746398,1.349632,0.633405,3.738736,2.924648
min,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,6.600000,0.000000,-1.012500
25%,1102.250000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,6.900000,8.000000,0.912150
50%,2203.500000,4.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,7.300000,11.000000,1.698150
75%,3304.750000,8.000000,1.000000,0.000000,0.000000,0.000000,0.000000,2.000000,7.800000,12.000000,3.172850
max,4406.000000,89.000000,104.000000,141.000000,155.000000,12.000000,8.000000,8.000000,10.900000,18.000000,54.835100



Head:


,Unnamed: 0,visits,nvisits,ovisits,novisits,emergency,hospital,health,chronic,adl,region,age,gender,married,school,income,employed,insurance,medicaid
0,1,5,0,0,0,0,1,average,2,normal,other,6.9,male,yes,6,2.8810,yes,yes,no
1,2,1,0,2,0,2,0,average,2,normal,other,7.4,female,yes,10,2.7478,no,yes,no
2,3,13,0,0,0,3,3,poor,4,limited,other,6.6,female,no,10,0.6532,no,no,yes
3,4,16,0,5,0,1,1,poor,2,limited,other,7.6,male,yes,3,0.6588,no,yes,no
4,5,3,0,0,0,0,0,average,2,limited,other,7.9,female,yes,6,0.6588,no,yes,no


## C1-T5 — Missing values + cleanliness check
**PDF requirement:** Determine if the data is clean / identify missing values.

### What I completed
- I calculated missing counts and percentages for every column.

### Results Capture
- Missing-count output shows zero null values across all columns.
- No missing-value treatment was required.

### Code evidence
- The next cell contains the exact missingness computation and table output.

In [6]:
# Missing values
na_counts = df.isna().sum().sort_values(ascending=False)
na_pct = (df.isna().mean() * 100).round(2)
missing_summary = pd.DataFrame({"missing_count": na_counts, "missing_pct": na_pct})
display(missing_summary[missing_summary["missing_count"] > 0] if (missing_summary["missing_count"] > 0).any() else missing_summary.head())


,missing_count,missing_pct
Unnamed: 0,0,0.0
visits,0,0.0
nvisits,0,0.0
ovisits,0,0.0
novisits,0,0.0


## C1-T6 — Comment on age + income (datatype + range + meaning)
**PDF requirement:** Comment on age and income: data type and values; determine range.

### What I completed
- I reviewed descriptive statistics for `age` and `income` and documented the encoded age interpretation.

### Results Capture
- `age` dtype=`float64`, range=`6.6` to `10.9` (years divided by 10).
- `income` dtype=`float64`, range=`-1.0125` to `54.8351`.
- Example interpretation: `age=6.9` corresponds to 69 years.

### Code evidence
- The next cell contains the column-level descriptive output used for this interpretation.

In [7]:
# Age + income notes
for col in ["age", "income"]:
    if col in df.columns:
        print(f"\n{col} dtype:", df[col].dtype)
        print(df[col].describe())
    else:
        print(f"Column not found: {col}")



age dtype: float64
count    4406.000000
mean        7.402406
std         0.633405
min         6.600000
25%         6.900000
50%         7.300000
75%         7.800000
max        10.900000
Name: age, dtype: float64

income dtype: float64
count    4406.000000
mean        2.527132
std         2.924648
min        -1.012500
25%         0.912150
50%         1.698150
75%         3.172850
max        54.835100
Name: income, dtype: float64


## C1-T7 — Export JSON and comment on format
**PDF requirement:** Export the dataframe to JSON (NSMES1988.json) and comment on the JSON.

### What I completed
- I exported the dataset to JSON with `records` orientation and previewed a snippet for format confirmation.

### Results Capture
- Export path: `outputs/NSMES1988.json`; orientation: `records`.
- The file is row-oriented and readable for downstream systems that expect object-per-record JSON.

### Artifacts
- `outputs/NSMES1988.json`

### Code evidence
- The next cell shows the export and snippet-preview code I executed.

In [8]:
# Export JSON
json_path = BASE_DIR / "outputs" / "NSMES1988.json"
json_path.parent.mkdir(parents=True, exist_ok=True)

df.to_json(json_path, orient="records")
print("Saved:", json_path)

# Preview first ~500 chars
with open(json_path, "r", encoding="utf-8") as f:
    snippet = f.read(500)
print("\nJSON snippet (first 500 chars):\n", snippet)

Saved: c:\DEV_Projects\SIMPLILEARN\CAPSTONE_Applied_Data_Science_with _Python\Incremental_Capstone\Capstone 1\outputs\NSMES1988.json

JSON snippet (first 500 chars):
 [{"Unnamed: 0":1,"visits":5,"nvisits":0,"ovisits":0,"novisits":0,"emergency":0,"hospital":1,"health":"average","chronic":2,"adl":"normal","region":"other","age":6.9,"gender":"male","married":"yes","school":6,"income":2.881,"employed":"yes","insurance":"yes","medicaid":"no"},{"Unnamed: 0":2,"visits":1,"nvisits":0,"ovisits":2,"novisits":0,"emergency":2,"hospital":0,"health":"average","chronic":2,"adl":"normal","region":"other","age":7.4,"gender":"female","married":"yes","school":10,"income":2.7478


## C1-T8 — Memory usage + dtype recommendations
**PDF requirement:** Provide memory information and recommend non-default data type changes.

### What I completed
- I measured total dataframe memory and identified safe dtype optimization recommendations.

### Results Capture
- Total memory usage: `2,263,919` bytes (`2.159 MB`).
- Recommended `category` candidates: `health`, `adl`, `region`, `gender`, `married`, `employed`, `insurance`, `medicaid`.

### Code evidence
- The next cell contains the memory computation and candidate dtype recommendation output.

In [9]:
# Memory usage
mem = df.memory_usage(deep=True).sum()
print("Total memory (bytes):", mem)
print("Total memory (MB):", round(mem / (1024**2), 3))
candidate_category = [c for c in ["health","adl","region","gender","married","employed","insurance","medicaid"] if c in df.columns]
print("Recommended category columns:", candidate_category)


Total memory (bytes): 2263919
Total memory (MB): 2.159
Recommended category columns: ['health', 'adl', 'region', 'gender', 'married', 'employed', 'insurance', 'medicaid']


## C1-T9 — Export cleaned CSV for next capstone
**PDF requirement:** Recommend dataframe changes and export cleaned dataframe as NSMES1988new.csv.

### What I completed
- I applied safe structural cleaning by removing the index-like column and exporting the cleaned dataframe.

### Results Capture
- Change applied: dropped `Unnamed: 0`.
- Export file: `outputs/NSMES1988new.csv`; resulting shape: `(4406, 18)`.

### Artifacts
- `outputs/NSMES1988new.csv`

### Code evidence
- The next cell contains the exact cleaning and export code I ran.

In [10]:
# Apply safe cleaning for capstone handoff
df_clean = df.copy()

if "Unnamed: 0" in df_clean.columns:
    df_clean = df_clean.drop(columns=["Unnamed: 0"])
    print("Dropped column: Unnamed: 0")

out_csv = BASE_DIR / "outputs" / "NSMES1988new.csv"
out_csv.parent.mkdir(parents=True, exist_ok=True)

df_clean.to_csv(out_csv, index=False)
print("Saved:", out_csv)
print("Shape:", df_clean.shape)

Dropped column: Unnamed: 0
Saved: c:\DEV_Projects\SIMPLILEARN\CAPSTONE_Applied_Data_Science_with _Python\Incremental_Capstone\Capstone 1\outputs\NSMES1988new.csv
Shape: (4406, 18)


## Final section — Conclusions (required)
- I completed the full Capstone 1 workflow for import, profiling, cleaning checks, and documentation.
- I confirmed the dataset is complete (no missing values) and documented age/income interpretation and ranges.
- I produced required artifacts: `outputs/NSMES1988.json` and `outputs/NSMES1988new.csv`.
- I updated `WORK_SUMMARY.md` with evidence and marked all tasks complete.


In [11]:
print("Capstone 1 completed: C1-T4 to C1-T9")
print("Artifacts: outputs/NSMES1988.json and outputs/NSMES1988new.csv")


Capstone 1 completed: C1-T4 to C1-T9
Artifacts: outputs/NSMES1988.json and outputs/NSMES1988new.csv
